File: 2026-01/2026-01-24/bingo_letter.py
def get_bingo_letter(n):

    return [bingo_range for bingo_range in [
        BingoRange("B", 1, 15),
        BingoRange("I", 16, 30),
        BingoRange("N", 31, 45),
        BingoRange("G", 46, 60),
        BingoRange("O", 61, 75)
    ] if bingo_range.min <= n <= bingo_range.max][0].letter

class BingoRange:
    def __init__(self, letter, min, max):
        self.letter = letter
        self.min = min
        self.max = max

# --- TEST SUITE ---

tests_text = r'''
1. get_bingo_letter(75) should return "O".
2. get_bingo_letter(54) should return "G".
3. get_bingo_letter(25) should return "I".
4. get_bingo_letter(38) should return "N".
5. get_bingo_letter(11) should return "B".
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

    print("----------------------------",
        f"\n⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "\n🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)

File: 01 January/01-24/bingoLetter.js
function getBingoLetter(n) {

  return [
    new BingoRange("B", 1, 15),
    new BingoRange("I", 16, 30),
    new BingoRange("N", 31, 45),
    new BingoRange("G", 46, 60),
    new BingoRange("O", 61, 75)
  ]
  .find(bingoRange => bingoRange.min <= n && n <= bingoRange.max)
  .letter;
}

class BingoRange {
  constructor(letter, min, max) {
    this.letter = letter;
    this.min = min;
    this.max = max;
  }
}

// --- TEST SUITE ---

const testsText = String.raw`
1. getBingoLetter(75) should return "O".
2. getBingoLetter(54) should return "G".
3. getBingoLetter(25) should return "I".
4. getBingoLetter(38) should return "N".
5. getBingoLetter(11) should return "B".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);

    const comparison = Array.isArray(testOutput)
      ? arraysEqual(functionCallOutput, testOutput)
      : functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      ++failCount;
    }
  })

  console.log("----------------------------",
    failCount
    ? `\n⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "\n🎉SUCCESS: All tests PASSED."
  );
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (let i = 0; i < a.length; ++i) {
    if (Array.isArray(a[i])) {
      if (Array.isArray(b[i])) {
        if (!arraysEqual(a[i], b[i]))
          return false;
      }
      else 
        return false;
    }

    else if (a[i] !== b[i])
      return false;
  }

  return true;
}

runTests(testData);